File: oxidize-golang/core/video/prompt.go
package video

import "fmt"

// PromptSegment is one block of a multimodal video prompt.
type PromptSegment struct {
	TextTokens []uint32
	Video      *VideoSegment
}

// VideoSegment holds per-frame embeddings flattened row-major.
type VideoSegment struct {
	Embeddings    []float32
	NumFrames     int
	LLMHiddenSize int
}

// VideoPrompt builds a flattened embedding sequence for video + text inputs.
type VideoPrompt struct {
	Segments              []PromptSegment
	VideoStartEmbedding   []float32
	VideoEndEmbedding     []float32
}

// NewVideoPrompt constructs an empty prompt.
func NewVideoPrompt() *VideoPrompt { return &VideoPrompt{} }

// AddText appends a text token block.
func (p *VideoPrompt) AddText(tokens []uint32) {
	p.Segments = append(p.Segments, PromptSegment{TextTokens: append([]uint32(nil), tokens...)})
}

// AddVideo appends a video embedding block.
func (p *VideoPrompt) AddVideo(embeddings []float32, numFrames, hidden int) {
	p.Segments = append(p.Segments, PromptSegment{
		Video: &VideoSegment{
			Embeddings:    append([]float32(nil), embeddings...),
			NumFrames:     numFrames,
			LLMHiddenSize: hidden,
		},
	})
}

// BuildSequence flattens segments using the token embedding table for text rows.
func (p *VideoPrompt) BuildSequence(table []float32, vocabSize, hiddenSize int) ([]float32, error) {
	llmHidden, err := p.inferHiddenSize(hiddenSize)
	if err != nil {
		return nil, err
	}
	totalRows, err := p.countRows(hiddenSize, llmHidden)
	if err != nil {
		return nil, err
	}
	out := make([]float32, totalRows*llmHidden)
	cursor := 0
	writeRow := func(row []float32) error {
		if len(row) != llmHidden {
			return &Error{Message: fmt.Sprintf("row width %d != %d", len(row), llmHidden)}
		}
		copy(out[cursor:cursor+llmHidden], row)
		cursor += llmHidden
		return nil
	}
	for _, seg := range p.Segments {
		if seg.Video != nil {
			if len(p.VideoStartEmbedding) == llmHidden {
				if err := writeRow(p.VideoStartEmbedding); err != nil {
					return nil, err
				}
			}
			v := seg.Video
			if v.NumFrames*v.LLMHiddenSize != len(v.Embeddings) {
				return nil, &Error{Message: "video embedding length mismatch"}
			}
			for f := 0; f < v.NumFrames; f++ {
				start := f * v.LLMHiddenSize
				if err := writeRow(v.Embeddings[start : start+v.LLMHiddenSize]); err != nil {
					return nil, err
				}
			}
			if len(p.VideoEndEmbedding) == llmHidden {
				if err := writeRow(p.VideoEndEmbedding); err != nil {
					return nil, err
				}
			}
			continue
		}
		for _, tok := range seg.TextTokens {
			if int(tok) >= vocabSize {
				return nil, &Error{Message: fmt.Sprintf("token %d >= vocab %d", tok, vocabSize)}
			}
			start := int(tok) * hiddenSize
			if start+hiddenSize > len(table) {
				return nil, &Error{Message: "embedding table too small"}
			}
			row := table[start : start+hiddenSize]
			if hiddenSize == llmHidden {
				if err := writeRow(row); err != nil {
					return nil, err
				}
				continue
			}
			padded := make([]float32, llmHidden)
			copy(padded, row)
			if err := writeRow(padded); err != nil {
				return nil, err
			}
		}
	}
	return out, nil
}

func (p *VideoPrompt) inferHiddenSize(fallback int) (int, error) {
	for _, seg := range p.Segments {
		if seg.Video != nil && seg.Video.LLMHiddenSize > 0 {
			return seg.Video.LLMHiddenSize, nil
		}
	}
	if fallback <= 0 {
		return 0, &Error{Message: "cannot infer hidden size"}
	}
	return fallback, nil
}

func (p *VideoPrompt) countRows(hiddenSize, llmHidden int) (int, error) {
	rows := 0
	for _, seg := range p.Segments {
		if seg.Video != nil {
			extra := 0
			if len(p.VideoStartEmbedding) == llmHidden {
				extra++
			}
			if len(p.VideoEndEmbedding) == llmHidden {
				extra++
			}
			rows += extra + seg.Video.NumFrames
			continue
		}
		rows += len(seg.TextTokens)
	}
	if rows == 0 {
		return 0, &Error{Message: "empty prompt"}
	}
	_ = hiddenSize
	return rows, nil
}

File: oxidize-python/oxidize_python/core/video/prompt.py
"""Multimodal video prompt mirroring oxidize-golang/core/video/prompt.go."""

from __future__ import annotations

from dataclasses import dataclass, field

from oxidize_python.core.video.errors import VideoError


@dataclass
class VideoSegment:
    """Per-frame embeddings flattened row-major."""

    embeddings: list[float]
    num_frames: int
    llm_hidden_size: int


@dataclass
class PromptSegment:
    """One block of a multimodal video prompt."""

    text_tokens: list[int] = field(default_factory=list)
    video: VideoSegment | None = None


class VideoPrompt:
    """Builds a flattened embedding sequence for video + text inputs."""

    def __init__(self) -> None:
        self.segments: list[PromptSegment] = []
        self.video_start_embedding: list[float] = []
        self.video_end_embedding: list[float] = []

    def add_text(self, tokens: list[int]) -> None:
        """Append a text token block."""
        self.segments.append(PromptSegment(text_tokens=list(tokens)))

    def add_video(self, embeddings: list[float], num_frames: int, hidden: int) -> None:
        """Append a video embedding block."""
        self.segments.append(
            PromptSegment(
                video=VideoSegment(
                    embeddings=list(embeddings),
                    num_frames=num_frames,
                    llm_hidden_size=hidden,
                )
            )
        )

    def build_sequence(
        self,
        table: list[float],
        vocab_size: int,
        hidden_size: int,
    ) -> list[float]:
        """Flatten segments using the token embedding table for text rows."""
        llm_hidden = self._infer_hidden_size(hidden_size)
        total_rows = self._count_rows(llm_hidden)
        out = [0.0] * (total_rows * llm_hidden)
        cursor = 0

        def write_row(row: list[float]) -> None:
            nonlocal cursor
            if len(row) != llm_hidden:
                raise VideoError(f"row width {len(row)} != {llm_hidden}")
            out[cursor : cursor + llm_hidden] = row
            cursor += llm_hidden

        for seg in self.segments:
            if seg.video is not None:
                if len(self.video_start_embedding) == llm_hidden:
                    write_row(self.video_start_embedding)
                v = seg.video
                if v.num_frames * v.llm_hidden_size != len(v.embeddings):
                    raise VideoError("video embedding length mismatch")
                for f in range(v.num_frames):
                    start = f * v.llm_hidden_size
                    write_row(v.embeddings[start : start + v.llm_hidden_size])
                if len(self.video_end_embedding) == llm_hidden:
                    write_row(self.video_end_embedding)
                continue
            for tok in seg.text_tokens:
                if tok >= vocab_size:
                    raise VideoError(f"token {tok} >= vocab {vocab_size}")
                start = tok * hidden_size
                if start + hidden_size > len(table):
                    raise VideoError("embedding table too small")
                row = table[start : start + hidden_size]
                if hidden_size == llm_hidden:
                    write_row(row)
                    continue
                padded = [0.0] * llm_hidden
                padded[: len(row)] = row[:llm_hidden]
                write_row(padded)
        return out

    def _infer_hidden_size(self, fallback: int) -> int:
        for seg in self.segments:
            if seg.video is not None and seg.video.llm_hidden_size > 0:
                return seg.video.llm_hidden_size
        if fallback <= 0:
            raise VideoError("cannot infer hidden size")
        return fallback

    def _count_rows(self, llm_hidden: int) -> int:
        rows = 0
        for seg in self.segments:
            if seg.video is not None:
                extra = 0
                if len(self.video_start_embedding) == llm_hidden:
                    extra += 1
                if len(self.video_end_embedding) == llm_hidden:
                    extra += 1
                rows += extra + seg.video.num_frames
                continue
            rows += len(seg.text_tokens)
        if rows == 0:
            raise VideoError("empty prompt")
        return rows
